Add an option to manage questionnaire access

bodhish opened the pull request:
This pull request includes significant refactoring to the routing and API request structure of the application, as well as the addition of a new component for managing questionnaire organizations.

### Routing and API Refactoring:

* [`src/Routers/AppRouter.tsx`](diffhunk://#diff-ba7691a84a8d897867e49026ab18bfd9b8899feb6dd37388d2c5e7989fcbf7a5L26-R27): Replaced direct imports of `QuestionnaireList` and `QuestionnaireShow` with a new `QuestionnaireRoutes` import and updated the `Routes` object to include `...QuestionnaireRoutes`. [[1]](diffhunk://#diff-ba7691a84a8d897867e49026ab18bfd9b8899feb6dd37388d2c5e7989fcbf7a5L26-R27) [[2]](diffhunk://#diff-ba7691a84a8d897867e49026ab18bfd9b8899feb6dd37388d2c5e7989fcbf7a5R57) [[3]](diffhunk://#diff-ba7691a84a8d897867e49026ab18bfd9b8899feb6dd37388d2c5e7989fcbf7a5L66-L68)
* [`src/Routers/routes/questionnaireRoutes.tsx`](diffhunk://#diff-a12a429ce597bc2187cd668065bb33c47d13a70cb6bc7f02697e9ea532fadec1R1-R11): Created a new file to define `QuestionnaireRoutes` for better modularity and separation of concerns.
* [`src/Utils/request/api.tsx`](diffhunk://#diff-ce6faa5d5e7bcf19601a5ec053a7a08e5a59a5193d0c65dc1ae383f460526745L52-L55): Removed old questionnaire and organization routes and updated the response types for user roles within organizations. [[1]](diffhunk://#diff-ce6faa5d5e7bcf19601a5ec053a7a08e5a59a5193d0c65dc1ae383f460526745L52-L55) [[2]](diffhunk://#diff-ce6faa5d5e7bcf19601a5ec053a7a08e5a59a5193d0c65dc1ae383f460526745L65) [[3]](diffhunk://#diff-ce6faa5d5e7bcf19601a5ec053a7a08e5a59a5193d0c65dc1ae383f460526745L610-L669) [[4]](diffhunk://#diff-ce6faa5d5e7bcf19601a5ec053a7a08e5a59a5193d0c65dc1ae383f460526745L748-L798) [[5]](diffhunk://#diff-ce6faa5d5e7bcf19601a5ec053a7a08e5a59a5193d0c65dc1ae383f460526745L819-R704)

### Component Updates:

* [`src/components/Questionnaire/ManageQuestionnaireOrganizationsSheet.tsx`](diffhunk://#diff-65e216f70ce8f7d5419c3e0162ffc5fa18fe9d4f783bc1050b4befc2cefd7836R1-R229): Added a new component to manage the organizations associated with a questionnaire, including the ability to add, remove, and search for organizations.

### API Usage Updates:

* Updated various components to use the new `questionnaireApi` and `organizationApi` modules instead of the old `routes` object:
  * `src/components/Questionnaire/QuestionnaireForm.tsx` [[1]](diffhunk://#diff-081d713c75e6bbb47f981cd70e404790762c4b9af5bdb5dd7b696523e6e30167R24) [[2]](diffhunk://#diff-081d713c75e6bbb47f981cd70e404790762c4b9af5bdb5dd7b696523e6e30167L73-R74)
  * `src/components/Questionnaire/QuestionnaireSearch.tsx` [[1]](diffhunk://#diff-1d17a521eaf4c9de6e397c76b0c82f8fa7aeb90062f059896c60ea8df0a1768dL14-R17) [[2]](diffhunk://#diff-1d17a521eaf4c9de6e397c76b0c82f8fa7aeb90062f059896c60ea8df0a1768dL41-R41)
  * `src/components/Questionnaire/index.tsx`
  * `src/components/Questionnaire/show.tsx` [[1]](diffhunk://#diff-cd36b7edf4b7da28f44bf14d9b5cea87a1bbe1c2bd300f7de6627033456afffcL12-R17) [[2]](diffhunk://#diff-cd36b7edf4b7da28f44bf14d9b5cea87a1bbe1c2bd300f7de6627033456afffcL69-R70) [[3]](diffhunk://#diff-cd36b7edf4b7da28f44bf14d9b5cea87a1bbe1c2bd300f7de6627033456afffcR110)
  * `src/pages/Landing/LandingPage.tsx` [[1]](diffhunk://#diff-180816d9f4530ff972e6e339dfae6ccaff2ce5070bf7cb16f629eb9ffa5ba9fdL16-R19) [[2]](diffhunk://#diff-180816d9f4530ff972e6e339dfae6ccaff2ce5070bf7cb16f629eb9ffa5ba9fdL35-R35)
  * `src/pages/Organization/OrganizationIndex.tsx`

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Added a new management interface for questionnaire organizations.
  - Introduced modular routing for questionnaire-related pages.

- **Improvements**
  - Streamlined API interactions across organization and questionnaire modules.
  - Enhanced type safety for API responses.
  - Centralized API endpoint definitions.

- **Technical Updates**
  - Refactored API route management.
  - Updated import and query mechanisms for organization and questionnaire data.
  - Simplified component data fetching logic.

- **Changes in API Structure**
  - Removed direct routes for questionnaires and organizations.
  - Created dedicated API modules for more organized data management.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

bodhish committed:
Refactor routing and API structure for organization and questionnaire components

- Moved questionnaire routes to a new file for better organization.
- Updated AppRouter to include the new QuestionnaireRoutes.
- Refactored organization API calls to use a new centralized organizationApi module, improving maintainability and clarity.
- Removed deprecated organization routes from the API definition.
- Updated various components to utilize the new organizationApi for API calls, ensuring consistency across the application.

bodhish committed:
Refactor questionnaire API routes and update components to use new structure

- Moved questionnaire routes from the api.tsx file to a new questionnaireApi module for better organization and maintainability.
- Updated components (QuestionnaireList, QuestionnaireForm, QuestionnaireSearch, and QuestionnaireShow) to utilize the new questionnaireApi for API calls, ensuring consistency across the application.
- Removed deprecated questionnaire routes from the previous API structure.

bodhish committed:
ManageQuestionnaireOrganizationsSheet component and integrate with QuestionnaireShow

- Introduced a new ManageQuestionnaireOrganizationsSheet component for managing organizations associated with a questionnaire.
- Integrated the new component into the QuestionnaireShow component, allowing users to add or remove organizations.
- Updated the questionnaireApi to reflect the correct request body structure for setting organizations.
- Enhanced the user interface for organization selection and management, including search functionality and visual feedback for selected organizations.

rithviknishad commented on src/components/Questionnaire/ManageQuestionnaireOrganizationsSheet.tsx:
```diff
@@ -0,0 +1,229 @@
+      mutate(questionnaireApi.setOrganizations, {
+        pathParams: { id: questionnaireId },
+        body: { organizations: organizations },
+      })({ organizations: organizations }),
```
```suggestion
    mutationFn: mutate(questionnaireApi.setOrganizations, {
        pathParams: { id: questionnaireId }
      }),
```

rithviknishad commented on src/components/Questionnaire/ManageQuestionnaireOrganizationsSheet.tsx:
```diff
@@ -0,0 +1,229 @@
+    selectedIds.includes(org.id),
+  );
+
+  const hasChanges =
+    JSON.stringify(organizations?.results.map((org) => org.id).sort()) !==
+    JSON.stringify(selectedIds.sort());
```
we can simply use `Set` for this instead right? not a fan of JSON stringify :)

rithviknishad commented on src/types/organization/organizationApi.ts:
```diff
@@ -0,0 +1,55 @@
+  removeUserRole: {
+    path: "/api/v1/organization/{id}/users/{userRoleId}/",
+    method: HttpMethod.DELETE,
+    TRes: {} as Record<string, never>,
+  },
+  listPatients: {
+    path: "/api/v1/patient/",
+    method: HttpMethod.GET,
+    TRes: Type<PaginatedResponse<Patient>>(),
```
for `TRes` can we stick with one of the formats?

```
TRes: Type<Record<string, never>>
// or
TRes: {} as Record<string, never>

bodhish committed:
Refactor ManageQuestionnaireOrganizationsSheet and organizationApi types

- Simplified the mutation function in ManageQuestionnaireOrganizationsSheet for setting organizations by removing redundant parameters.
- Updated the handleSave function to correctly format the payload for the setOrganizations mutation.
- Enhanced the hasChanges logic to improve clarity and accuracy in detecting changes in selected organizations.
- Refactored organizationApi type definitions to use Type<> for response and request bodies, improving type safety and consistency across API calls.
